k2/csrc/hash.h

```diff
@@ -1014,6 +1014,349 @@ class Hash {
+        uint64_t *key_value_location, uint64_t key, uint64_t value) const {
+      *key_value_location = key;
+      *(key_value_location + 1) = value;
+    }
+
+    /* Deletes a key from a hash.  Caution: this cannot be combined with other
+       operations on a hash; after you delete a key you cannot do Insert() or
+       Find() until you have deleted all keys.  This is an open-addressing hash
+       table with no tombstones, which is why this limitation exists).
+
+       @param [in] key   Key to be deleted.   Each key present in the hash must
+                         be deleted  by exactly one thread, or it will loop
+                         forever!
+
+      Note: the const is with respect to the metadata only; required, to avoid
+      compilation errors.
+    */
+    __forceinline__ __host__ __device__ void Delete(uint64_t key) const {
+      uint32_t cur_bucket = key & num_buckets_mask_,
+               bucket_inc = 1 | ((key >> buckets_num_bitsm1_) ^ key);
+      while (1) {
+        uint64_t old_key = data_[2 * cur_bucket];
+        if (old_key == key) {
+          data_[2 * cur_bucket] = ~((uint64_t)0);
+          data_[2 * cur_bucket + 1] = ~((uint64_t)0);
+          return;
+        } else {
+          cur_bucket = (cur_bucket + bucket_inc) & num_buckets_mask_;
+        }
+      }
+    }
+
+   private:
+    // pointer to data
+    uint64_t *data_;
+    // num_buckets_mask is num_buckets (i.e. size of `data_` array) minus one;
+    // num_buckets is a power of 2 so this can be used as a mask to get a number
+    // modulo num_buckets.
+    uint32_t num_buckets_mask_;
+    // A number satisfying num_buckets == 1 << (1+buckets_num_bitsm1_)
+    // the number of bits in `num_buckets` minus one.
+    uint32_t buckets_num_bitsm1_;
+  };
+
+  /*
+    Return an Accessor object which can be used in kernel code (or on CPU if the
+    context is a CPU context).
+  */
+  Accessor GetAccessor() { return Accessor(*this); }
+
+  // You should call this before the destructor is called if the hash will still
+  // contain values when it is destroyed, to bypass a check.
+  void Destroy() { data_ = Array1<uint64_t>(); }
+
+  void CheckEmpty() {
+    if (data_.Dim() == 0) return;
+    ContextPtr c = Context();
+    Array1<int32_t> error(c, 1, -1);
+    int32_t *error_data = error.Data();
+    uint64_t *hash_data = data_.Data();
+
+    K2_EVAL(
+        Context(), data_.Dim(), lambda_check_data, (int32_t i)->void {
+          if (~(hash_data[i]) != 0) error_data[0] = i;
+        });
+    int32_t i = error[0];
+    if (i >= 0) {  // there was an error; i is the index into the hash where
+      // there was an element.
+      int64_t elem = data_[i];
+      // We don't know the number of bits the user was using for the key vs.
+      // value, so print in hex, maybe they can figure it out.
+      K2_LOG(FATAL) << "Destroying hash: still contains values: position " << i
+                    << ", key,value = " << std::hex << elem;
```

Comment:
```suggestion
                    << ", content = " << std::hex << elem;
```
I think `elem` can be either a key or a value.